### PR TITLE
[BEAM-5203] expose PaneInfo and BoundedWindow as UDF

### PR DESCRIPTION
**Pull request by XuMingmin**
Add UDFs to expose `BoundedWindow` and `PaneInfo` details. And it's also a blocker for (BEAM-5204: reduce restriction of JOIN). Now JOIN is acceptable only when the upstream is triggered once per window;

Note that, this is the solution using annotation `UDF`. Another potential way is adding FUNCTION keywords in Calcite-core, it may not be a good idea as these are Beam-specific concepts, especially `PaneInfo`.

Changes in summary:
1. put `BoundedWindow`  and `PaneInfo` in `BeamSqlExpressionEnvironment`;
2. create UDFs in `PaneInfoUdfs` and `WindowUdfs`;

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by XuMingmin** [comment]
@apilloud @akedin @amaliujia
hello, can anyone take a look?

**Comment by apilloud** [comment]
(cc: @kennknowles  This has SQL language design implications you should be aware of.)

Sorry for missing this the first time, I'll spend the morning looking into it. It looks like the actual goal of this is not to add a windowed UDF but to add 7 built in functions for using in joins: FIRST_PANE, LAST_PANE, PANE_TIMING, PANE_INDEX, WINDOW_TYPE, WINDOW_START, WINDOW_END

There are performance benefits to not running UDFs in window context. I have a outstanding PR that removes windows from UDFs entirely: #6417 That same PR deletes most of this code. I think adding the functions directly would be more compatible with that and could probably be done without changing calcite, so let me do some digging.

**Comment by kennknowles** [comment]
This level of fundamental change should definitely be discussed on the dev list. Was it? I am still getting through my email after a vacation so I may not have seen the thread.

**Comment by XuMingmin** [comment]
Thanks @apilloud , sounds like it would conflict with #6417 as Window/Pane is not accessible now, let me take a look at that PR. 

The goal for this PR is to expose Beam Window/Pane details via UDFs. It may be useful in more scenarios, while the motivation from my side is remove the strict limitation on JOIN.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/udf/WindowUdfs.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    return "UNKNOWN";
+  }
+
+  @UDF(
+    funcName = "WINDOW_START",
+    parameterArray = {},
+    returnType = Schema.TypeName.DATETIME
+  )
+  public Instant windowStartTime() {
+    return Instant.now();
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/udf/WindowUdfs.java]: Is it intentional to return Instant.now?

**Comment by XuMingmin** [comment]
Close it for further option as it conflicts with #6417

**Pull request closed without merging**